ledger/complete/mtrie/flattener/iterator.go

```diff
@@ -115,7 +114,7 @@ func (i *NodeIterator) Next() bool {
 		// initial call to Next() for a non-empty trie
 		i.dig(i.unprocessedRoot)
 		i.unprocessedRoot = nil
-		return true
+		return len(i.stack) > 0
```

Comment:
This is a bug fix for a problem that didn't surface yet because of the way node iterator was used.  

The bug is unique node iterator's `Next()` returns true when `i.unprocessedRoot` is visited already and `i.stack` is empty.

This bug doesn't happen when we iterate nodes of an entire trie (root nodes are always unique).  In this PR, we iterate nodes of subtries and subtries can be shared and visited already.  So instead of always returning `true` assuming there's at least one unique node when digging `i.unprocessedRoot`, we only return true when there are unique nodes in the internal stack after calling `dig(i.unprocessedRoot)`.

Reply:
Thanks for the explanation! It does make sense.
Would it be possible to maybe add a test catching this particular bug, and showing how the fix helps?

Reply:
> Would it be possible to maybe add a test catching this particular bug, and showing how the fix helps?

Test for this bug is already in [iterator_test.go#L269-L396](https://github.com/onflow/flow-go/blob/d136a3e1e795af8606e1e53a67ee79276c799705/ledger/complete/mtrie/flattener/iterator_test.go#L269-L396).

The test iterates 3 left substries and 3 right subtries (some subtries are shared).  The test verifies that:
- order of iterated nodes is descendents first
- shared subtries/nodes are not iterated twice
- non-nil node is returned (meaning as long as `Next()` returns true `Value()` returns a non-nil node)